src/io/ioexternalbus.cpp

```diff
@@ -28,10 +45,201 @@ QString IOExternalBus::description() const {
 }
 
 VInt IOExternalBus::ioRead(AInt offset, unsigned size) {
-    return 0;
+    uint32_t rvalue = 0;
+
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload = htonl(offset);
+        if (send_cmd(VB_PREAD, 4, (const char*)&payload) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(&cmd_header) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        if (cmd_header.payload_size) {
+            uint32_t* payload = new uint32_t[2];
+            if (recv_payload((char*)payload, 8) < 0) {
+                skt_use = 0;
+                return 0;
+            }
+            for (uint32_t i = 0; i < 2; i++) {
+                payload[i] = ntohl(payload[i]);
+            }
+            dprintf("Read addr[%x] = %x \n", payload[0], payload[1]);
+            rvalue = payload[1];
+            delete[] payload;
+        } else {
+            skt_use = 0;
+            printf("read error [%lx] (%x) msg=%i pyaload_size=%i !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n", offset, size,
+                   cmd_header.msg_type, cmd_header.payload_size);
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+
+    dprintf("===============>ioRead [%lx] = %x (%x)\n", offset, rvalue, size);
+    return rvalue;
+}
+
+void IOExternalBus::ioWrite(AInt offset, VInt value, unsigned size) {
+    dprintf("===============>ioWrite [%lx] = %lx (%x)\n", offset, value, size);
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload[2];
+        payload[0] = htonl(offset);
+        payload[1] = htonl(value);
+        if (send_cmd(VB_PWRITE, 8, (const char*)payload) < 0) {
+            skt_use = 0;
+            return;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(&cmd_header) < 0) {
+            skt_use = 0;
+            return;
+        }
+
+        if (cmd_header.msg_type == VB_PWRITE) {
+            dprintf("write ok\n");
+        } else {
+            skt_use = 0;
+            printf("write error !!!!!!!!!!!!!!!!!!!!!!\n");
```

Comment:
Could be sent to a ui component.

Reply:
Error messages are currently shown with a `QMessageBox::information` case the error occur in a different IO tab.